service/controller/v5/resource/instance/deployment.go

```diff
@@ -35,34 +67,70 @@ func (r Resource) newDeployment(ctx context.Context, obj providerv1alpha1.AzureC
 		workerNodes = append(workerNodes, n)
 	}
 
-	cc, err := controllercontext.FromContext(ctx)
+	containerName := key.BlobContainerName()
+	groupName := key.ResourceGroupName(obj)
+	storageAccountName := key.StorageAccountName(obj)
+	masterBlobName := key.BlobName(obj, key.PrefixMaster())
+	workerBlobName := key.BlobName(obj, key.PrefixWorker())
+	encryptionKey := cc.CloudConfig.GetEncryptionKey()
+	initialVector := cc.CloudConfig.GetInitialVector()
+
+	storageAccountsClient, err := r.getStorageAccountsClient(ctx)
 	if err != nil {
 		return azureresource.Deployment{}, microerror.Mask(err)
 	}
-	err = cc.Validate()
+
+	keys, err := storageAccountsClient.ListKeys(ctx, groupName, storageAccountName)
 	if err != nil {
 		return azureresource.Deployment{}, microerror.Mask(err)
 	}
 
-	masterCloudConfig, err := cc.CloudConfig.NewMasterCloudConfig(obj)
+	if len(*(keys.Keys)) == 0 {
+		return azureresource.Deployment{}, microerror.Maskf(executionFailedError, "storage account key's list is empty")
+	}
+	primaryKey := *(((*keys.Keys)[0]).Value)
+
+	masterBlobURL, err := blobclient.GetBlobURL(ctx, masterBlobName, containerName, storageAccountName, primaryKey, cc.ContainerURL)
 	if err != nil {
 		return azureresource.Deployment{}, microerror.Mask(err)
 	}
 
-	workerCloudConfig, err := cc.CloudConfig.NewWorkerCloudConfig(obj)
+	c := SmallCloudconfigConfig{
+		BlobURL:       masterBlobURL,
+		EncryptionKey: encryptionKey,
+		InitialVector: initialVector,
+	}
+	masterCloudConfig, err := templates.Render(key.CloudConfigSmallTemplates(), c)
+	if err != nil {
+		return azureresource.Deployment{}, microerror.Mask(err)
+	}
+	encodedMasterCloudConfig := base64.StdEncoding.EncodeToString([]byte(masterCloudConfig))
```

Comment:
Why do we have to encode the rendered template all the time? What about letting the rendering do this? Are there exceptions where we do not want the encoding to happen?

Reply:
There are different encodings: files encoding for ignition and final config ignition for VMSS user-data